cpp/src/arrow/filesystem/s3fs_test.cc

```diff
@@ -1005,7 +1005,8 @@ TEST_F(TestS3FS, CreateDir) {
   AssertFileInfo(fs_.get(), "bucket/newdir/newsub/newsubsub", FileType::Directory);
 
   // Existing "file", should fail
-  ASSERT_RAISES(IOError, fs_->CreateDir("bucket/somefile"));
+  ASSERT_RAISES(IOError, fs_->CreateDir("bucket/somefile", /*recursive=*/false));
+  ASSERT_RAISES(IOError, fs_->CreateDir("bucket/somefile", /*recursive=*/true));
```

Comment:
Someday the filesystem tests could be unified. I was very careful with directory semantics in the azure fs implementations. Testing all cases to ensure they behave like Linux system calls.

Reply:
Yes, we could open a new issue to improve the generic filesystem tests. Would you like to do that?

Reply:
> Would you like to do that?

No. Because I know that the other filesystems diverge and these details have leaked into user code (i.e. they rely on the broken semantics).

Reply:
Fair enough. Usually there's a tension into enforcing consistent semantics accross filesystems, and trying to preserve performance when consistency implies more IO roundtrips to a remote host. If there are inconsistencies that are not motivated by performance, they should be investigated.

Reply:
S3 finally supports Conditional Writes. That means a lot of the check-before-create sequences you need to implement the filesystem semantics can be made much faster now.

https://aws.amazon.com/about-aws/whats-new/2024/08/amazon-s3-conditional-writes/

Reply:
Ok, I opened https://github.com/apache/arrow/issues/44281 for it